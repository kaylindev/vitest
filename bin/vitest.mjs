#!/usr/bin/env node
'use strict'

import { fileURLToPath } from 'url'
import { resolve, dirname } from 'path'
import { startAndRun } from 'vite-node'
import minimist from 'minimist'

const argv = minimist(process.argv.slice(2), {
  string: ['root'],
})

const __dirname = dirname(fileURLToPath(import.meta.url))

await startAndRun({
  root: resolve(argv.root || process.cwd()),
  files: [
    resolve(__dirname, '../dist/cli.js'),
  ],
})
