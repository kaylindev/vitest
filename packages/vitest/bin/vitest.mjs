#!/usr/bin/env node

import { fileURLToPath } from 'url'
import { resolve } from 'pathe'
import { ensurePackageInstalled } from '../dist/utils.js'

const argv = process.argv.slice(2)

if (!await ensurePackageInstalled('vite'))
  process.exit(1)

if (argv.includes('--coverage')) {
  if (!await ensurePackageInstalled('c8'))
    process.exit(1)
  const filename = fileURLToPath(import.meta.url)
  const entry = resolve(filename, '../../dist/cli.js')
  process.argv.splice(2, 0, process.argv[0], entry)
  await import('c8/bin/c8.js')
}
else {
  await import('../dist/cli.js')
}
