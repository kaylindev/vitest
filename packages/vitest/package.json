{
  "name": "vitest",
  "version": "0.0.92",
  "description": "A blazing fast unit test framework powered by Vite",
  "keywords": [
    "vite",
    "vite-node",
    "test",
    "jest"
  ],
  "homepage": "https://github.com/antfu/vitest#readme",
  "bugs": {
    "url": "https://github.com/antfu/vitest/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/antfu/vitest.git"
  },
  "funding": "https://github.com/sponsors/antfu",
  "license": "MIT",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./node": {
      "import": "./dist/node.js",
      "types": "./dist/node.d.ts"
    },
    "./global": {
      "types": "./global.d.ts"
    },
    "./*": "./*"
  },
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "bin": {
    "vitest": "./bin/vitest.mjs"
  },
  "files": [
    "dist",
    "bin",
    "*.d.ts"
  ],
  "scripts": {
    "build": "rimraf dist && rollup -c",
    "dev": "rollup -c --watch src",
    "typecheck": "tsc --noEmit",
    "prepublishOnly": "nr build"
  },
  "dependencies": {
    "@types/chai": "^4.3.0",
    "@types/chai-subset": "^1.3.3",
    "chai": "^4.3.4",
    "chai-subset": "^1.6.0",
    "fast-glob": "^3.2.7",
    "local-pkg": "^0.4.0",
    "piscina": "^3.2.0",
    "source-map": "^0.7.3",
    "tinyspy": "^0.1.0"
  },
  "devDependencies": {
    "@antfu/install-pkg": "^0.1.0",
    "@types/diff": "^5.0.1",
    "@types/jsdom": "^16.2.13",
    "@types/micromatch": "^4.0.2",
    "@types/natural-compare": "^1.4.1",
    "@types/node": "^17.0.0",
    "@types/prompts": "^2.0.14",
    "c8": "^7.10.0",
    "cac": "^6.7.12",
    "cli-truncate": "^3.1.0",
    "diff": "^5.0.0",
    "find-up": "^6.2.0",
    "flatted": "^3.2.4",
    "happy-dom": "^2.24.5",
    "jsdom": "^19.0.0",
    "log-update": "^5.0.0",
    "micromatch": "^4.0.4",
    "mlly": "^0.3.15",
    "nanoid": "^3.1.30",
    "natural-compare": "^1.4.0",
    "pathe": "^0.2.0",
    "picocolors": "^1.0.0",
    "pkg-types": "^0.3.2",
    "pretty-format": "^27.4.2",
    "prompts": "^2.4.2",
    "rollup": "^2.61.1",
    "strip-ansi": "^7.0.1",
    "typescript": "^4.5.4"
  },
  "peerDependencies": {
    "c8": "*",
    "happy-dom": "*",
    "jsdom": "*",
    "vite": "^2.7.1"
  },
  "peerDependenciesMeta": {
    "c8": {
      "optional": true
    },
    "happy-dom": {
      "optional": true
    },
    "jsdom": {
      "optional": true
    }
  },
  "engines": {
    "node": ">=16.0.0"
  }
}
