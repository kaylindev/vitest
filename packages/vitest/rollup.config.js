import fs from 'fs'
import path from 'pathe'
import esbuild from 'rollup-plugin-esbuild'
import dts from 'rollup-plugin-dts'
import resolve from '@rollup/plugin-node-resolve'
import commonjs from '@rollup/plugin-commonjs'
import json from '@rollup/plugin-json'
import alias from '@rollup/plugin-alias'
import license from 'rollup-plugin-license'
import c from 'picocolors'
import fg from 'fast-glob'
import pkg from './package.json'

const entry = [
  'src/index.ts',
  'src/node/cli.ts',
  'src/node.ts',
  'src/runtime/worker.ts',
  'src/runtime/entry.ts',
  'src/utils.ts',
]

const external = [
  ...Object.keys(pkg.dependencies),
  ...Object.keys(pkg.peerDependencies),
  'worker_threads',
]

export default ({ watch }) => [
  {
    input: entry,
    output: {
      dir: 'dist',
      format: 'esm',
    },
    external,
    plugins: [
      alias({
        entries: [
          { find: /^node:(.+)$/, replacement: '$1' },
        ],
      }),
      resolve({
        preferBuiltins: true,
      }),
      json(),
      commonjs(),
      esbuild({
        target: 'node14',
      }),
      !watch && licensePlugin(),
    ],
    onwarn(message) {
      if (/Circular dependencies/.test(message))
        return
      console.error(message)
    },
  },
  {
    input: [
      'src/index.ts',
    ],
    output: {
      file: 'dist/index.d.ts',
      format: 'esm',
    },
    external,
    plugins: [
      dts(),
    ],
  },
  {
    input: [
      'src/node.ts',
    ],
    output: {
      file: 'dist/node.d.ts',
      format: 'esm',
    },
    external,
    plugins: [
      dts(),
    ],
  },
]

function licensePlugin() {
  return license({
    thirdParty(dependencies) {
      // https://github.com/rollup/rollup/blob/master/build-plugins/generate-license-file.js
      // MIT Licensed https://github.com/rollup/rollup/blob/master/LICENSE-CORE.md
      const coreLicense = fs.readFileSync(
        path.resolve(__dirname, '../../LICENSE'),
      )
      function sortLicenses(licenses) {
        let withParenthesis = []
        let noParenthesis = []
        licenses.forEach((license) => {
          if (/^\(/.test(license))
            withParenthesis.push(license)

          else
            noParenthesis.push(license)
        })
        withParenthesis = withParenthesis.sort()
        noParenthesis = noParenthesis.sort()
        return [...noParenthesis, ...withParenthesis]
      }
      const licenses = new Set()
      const dependencyLicenseTexts = dependencies
        .sort(({ name: nameA }, { name: nameB }) =>
          nameA > nameB ? 1 : nameB > nameA ? -1 : 0,
        )
        .map(
          ({
            name,
            license,
            licenseText,
            author,
            maintainers,
            contributors,
            repository,
          }) => {
            let text = `## ${name}\n`
            if (license)
              text += `License: ${license}\n`

            const names = new Set()
            if (author && author.name)
              names.add(author.name)

            for (const person of maintainers.concat(contributors)) {
              if (person && person.name)
                names.add(person.name)
            }
            if (names.size > 0)
              text += `By: ${Array.from(names).join(', ')}\n`

            if (repository)
              text += `Repository: ${repository.url || repository}\n`

            if (!licenseText) {
              try {
                const pkgDir = path.dirname(
                  resolve(path.join(name, 'package.json'), {
                    preserveSymlinks: false,
                  }),
                )
                const licenseFile = fg.sync(`${pkgDir}/LICENSE*`, {
                  caseSensitiveMatch: false,
                })[0]
                if (licenseFile)
                  licenseText = fs.readFileSync(licenseFile, 'utf-8')
              }
              catch {}
            }
            if (licenseText) {
              text
                += `\n${
                  licenseText
                    .trim()
                    .replace(/(\r\n|\r)/gm, '\n')
                    .split('\n')
                    .map(line => `> ${line}`)
                    .join('\n')
                }\n`
            }
            licenses.add(license)
            return text
          },
        )
        .join('\n---------------------------------------\n\n')
      const licenseText
        = '# Vitest core license\n'
        + `Vitest is released under the MIT license:\n\n${
          coreLicense
        }\n# Licenses of bundled dependencies\n`
        + 'The published Vitest artifact additionally contains code with the following licenses:\n'
        + `${sortLicenses(licenses).join(', ')}\n\n`
        + `# Bundled dependencies:\n${
          dependencyLicenseTexts}`
      const existingLicenseText = fs.readFileSync('LICENSE.md', 'utf8')
      if (existingLicenseText !== licenseText) {
        fs.writeFileSync('LICENSE.md', licenseText)
        console.warn(
          c.yellow(
            '\nLICENSE.md updated. You should commit the updated file.\n',
          ),
        )
      }
    },
  })
}
