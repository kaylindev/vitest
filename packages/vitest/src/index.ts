import type { Anything, ArrayContaining } from './integrations/chai/jest-asymmetric-matchers'
import type { MatchersObject } from './integrations/chai/types'
import type { InlineConfig as VitestInlineConfig } from './types'

export { suite, test, describe, it } from './runtime/suite'
export * from './types'
export * from './runtime/hooks'
export * from './integrations/chai'
export * from './integrations/tinymock'
export * from './integrations/utils'

declare module 'vite' {
  interface UserConfig {
    /**
     * Options for Vitest
     */
    test?: VitestInlineConfig
  }
}

declare global {
  namespace Chai {
    interface ExpectStatic {
      extend(expects: MatchersObject): void
      stringContaining(expected: string): void
      anything(): Anything
      arrayContaining(expected: any): ArrayContaining
    }

    interface Assertion {
      // Chai
      chaiEqual(expected: any): void

      // Snapshot
      toMatchSnapshot(message?: string): Assertion
      toMatchInlineSnapshot(snapshot?: string, message?: string): Assertion
      matchSnapshot(message?: string): Assertion

      // Jest compact
      toEqual(expected: any): void
      toStrictEqual(expected: any): void
      toBe(expected: any): void
      toMatch(expected: string | RegExp): void
      toMatchObject(expected: any): void
      toContain(item: any): void
      toContainEqual(item: any): void
      toBeTruthy(): void
      toBeFalsy(): void
      toBeGreaterThan(num: number): void
      toBeGreaterThanOrEqual(num: number): void
      toBeLessThan(num: number): void
      toBeLessThanOrEqual(num: number): void
      toBeNaN(): void
      toBeUndefined(): void
      toBeNull(): void
      toBeDefined(): void
      toBeInstanceOf(c: any): void
      toBeCalledTimes(n: number): void
      toHaveLength(l: number): void
      toHaveProperty(p: string, value?: any): void
      toBeCloseTo(number: number, numDigits?: number): void
      toHaveBeenCalledTimes(n: number): void
      toHaveBeenCalledOnce(): void
      toHaveBeenCalled(): void
      toBeCalled(): void
      toHaveBeenCalledWith(...args: any[]): void
      toBeCalledWith(...args: any[]): void
      toHaveBeenNthCalledWith(n: number, ...args: any[]): void
      nthCalledWith(n: number, ...args: any[]): void
      toHaveBeenLastCalledWith(...args: any[]): void
      lastCalledWith(...args: any[]): void
      toThrow(expected?: string | RegExp): void
      toThrowError(expected?: string | RegExp): void
      toReturn(): void
      toHaveReturned(): void
      toReturnTimes(times: number): void
      toHaveReturnedTimes(times: number): void
      toReturnWith(value: any): void
      toHaveReturnedWith(value: any): void
      toHaveLastReturnedWith(value: any): void
      lastReturnedWith(value: any): void
      toHaveNthReturnedWith(nthCall: number, value: any): void
      nthReturnedWith(nthCall: number, value: any): void
    }
  }
}
