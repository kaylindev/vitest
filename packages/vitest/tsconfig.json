{
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "lib": ["esnext", "dom"],
    "moduleResolution": "node",
    "esModuleInterop": true,
    "strict": true,
    "strictNullChecks": true,
    "resolveJsonModule": true,
    "skipDefaultLibCheck": true,
    "skipLibCheck": true,
    "outDir": "./dist",
    "declaration": true,
    "inlineSourceMap": true,
    "paths": {
      "vitest": ["./packages/vitest/src/index.ts"]
    }
  },
  "include": [
    "./src/**/*.ts"
  ],
  "exclude": [
    "./dist"
  ]
}
