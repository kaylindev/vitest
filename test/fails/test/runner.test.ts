import { resolve } from 'path'
import fg from 'fast-glob'
import { execa } from 'execa'
import { describe, it, expect } from 'vitest'

describe('should fails', async() => {
  const root = resolve(__dirname, '../fixtures')
  const cli = resolve(__dirname, '../../../bin/vitest.mjs')
  const files = await fg('*.test.ts', { cwd: root })

  for (const file of files) {
    it(file, async() => {
      let error: any
      await execa('node', [cli, file], {
        cwd: root,
        env: {
          ...process.env,
          CI: 'true',
          NO_COLOR: 'true',
        },
      })
        .catch((e) => {
          error = e
        })

      expect(error).toBeTruthy()
      const msg = String(error)
        .split(/\n/g)
        .reverse()
        .find(i => i.includes('Error: '))
        ?.trim()
      expect(msg).toMatchSnapshot(file)
    }, 10000)
  }
})
