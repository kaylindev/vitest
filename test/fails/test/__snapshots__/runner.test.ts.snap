// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fails > expect.test.ts 1`] = `"AssertionError: expected 2 to deeply equal 3"`;

exports[`should fails > nested-suite.test.ts 1`] = `"AssertionError: expected true to equal false"`;
